Revamp the storage quota and eviction documentation (#25019)
mdn/content: files/en-us/learn/javascript/client-side_web_apis/browser_storage_limits_and_eviction_criteria/index.md

---
title: Browser storage limits and eviction criteria

> slug: >-
>   Learn/JavaScript/Client-side_web_APIs/Browser_storage_limits_and_eviction_criteria

Reviewer: ```suggestion
slug: Learn/JavaScript/Client-side_web_APIs/Browser_storage_limits_and_eviction_criteria
```

Author: Not sure what I should do here. This change was created automatically when I used the redirect command.

This term is important in this article because browsers manage stored data _per origin_. The quotas and eviction criteria described in this article apply to an entire origin, even if this origin is used to run several websites, such as `https://example.com/site1/` and `https://example.com/site2/`.

## What technologies store data in the browser?

Web developers can use the following web technologies to store data in the browser:

| Technology | Description |
| ---------- | ----------- |
| [Cookies](/en-US/docs/Web/HTTP/Cookies) | An HTTP cookie is a small piece of data that the web server and browser send each other to remember stateful information across page navigation. |
| [Web Storage](/en-US/docs/Web/API/Web_Storage_API) | The Web Storage API provides mechanisms for webpages to store string-only key/value pairs. |
| [IndexedDB](/en-US/docs/Web/API/IndexedDB_API) | IndexedDB is a Web API for storing large data structures in the browser and indexing them for high-performance searching. |
| [Cache API](/en-US/docs/Web/API/Cache) | The Cache API provides a persistent storage mechanism for HTTP request and response object pairs that's used to make webpages load faster. |

> | [Origin Private File System Access API (OPFS)](https://developer.chrome.com/articles/file-system-access/#accessing-the-origin-private-file-system) | OPFS provides a file system that's private to the origin of the page and can be used to read and write directories and files. |

Author: Yes, that's a good idea, I'll create an issue.

> | [Origin Private File System Access API (OPFS)](https://developer.chrome.com/articles/file-system-access/#accessing-the-origin-private-file-system) | OPFS provides a file system that's private to the origin of the page and can be used to read and write directories and files. |

Author: Oh, we already have OPFS documented! https://developer.mozilla.org/en-US/docs/Web/API/File_System_Access_API#origin_private_file_system
I'll update the link here.

## Does browser-stored data persist?

Data for an origin can be stored in two ways in a browser, _persistent_ and _best-effort_:

- Best-effort: this is the way that data is stored by default. Best-effort data persists as long as the origin is below its quota, the device has enough storage space, and the user doesn't choose to delete the data via their browser's settings.
- Persistent: an origin can opt-in to store its data in a persistent way. Data stored this way is only evicted if the user chooses to, by using their browser's settings.

The data stored in the browser by a origin is best-effort by default. When using web technologies such as IndexedDB or Cache, the data is stored transparently without asking for the user's permission. Similarly, when the browser needs to evict best-effort data, it does so without interrupting the user.

> If, for any reasons, developers need persistent storage (e.g., when building a web app that relies on critical data that isn't persisted anywhere else), they can do so by using the {{domxref("StorageManager.persist()")}} method of the {{domxref("Storage_API", "Storage API", "", "nocode")}}.

Reviewer 2: ```suggestion
If, for any reason, developers need persistent storage (e.g., when building a web app that relies on critical data that isn't persisted anywhere else), they can do so by using the {{domxref("StorageManager.persist()")}} method of the {{domxref("Storage_API", "Storage API", "", "nocode")}}.
```

> In Firefox, when a site chooses to use persistent storage, the user is notified with a UI popup their permission is requested.

Reviewer 2: ```suggestion
In Firefox, when a site chooses to use persistent storage, the user is notified with a UI popup that their permission is requested.
```

Safari and most Chromium-based browsers, such as Chrome or Edge, automatically handle the permission request, and do not show any prompts to the user.

> Note that research from the Chrome team shows that data is very rarely deleted by the browser (source: [Persistent storage](https://web.dev/persistent-storage/) at web.dev). If a user uses a website regularly, there are very little chances that its stored data, even in best-effort mode, will get evicted by the browser.

Reviewer 2: ```suggestion
Note that research from the Chrome team shows that data is very rarely deleted by the browser (source: [Persistent storage](https://web.dev/persistent-storage/) at web.dev). If a user uses a website regularly, there is very little chance that its stored data, even in best-effort mode, will get evicted by the browser.
```

### Private browsing

> In private browsing mode (also called _Incognito_ in Chrome, and _InPrivate_ in Edge), stored data is deleted when the private window is closed.

Author: I don't know for sure. It does seem to be shared in my quick test in Edge. Not sure if all browsers work the same way. It's probably best to keep this note high-level enough to apply to all browsers. The point here is that data does not persist in this mode. I'll attempt something.

This eviction mechanism only applies to origins that are not persistent, and skips over origins that have requested the user for persistent data by using {{domxref("StorageManager.persist()")}}).

### Browser maximum storage exceeded eviction

Some browsers define a maximum storage space that they can use on the device's hard disk. For example, Chrome currently uses at most 80% of the total disk size.

This maximum storage size means that there may come a point at which the data stored by all of the combined origins exceeds the maximum size without any one origin being above its individual quota.

When this happens, the browser starts evicting best-effort origins as described in [Storage pressure eviction](#storage-pressure-eviction).

### Proactive eviction

Currently, Safari is the only browser that proactively evicts data. The data from any origin that hasn't been accessed by the user over the last seven days is deleted.

## How is data evicted?

> When an origin's data is evicted by the browser, all of its data, not parts of it, is deleted at the same time. If the origin had stored data by using IndexedDB and the Cache API for example, then both types of data is deleted.

Reviewer: ```suggestion
When an origin's data is evicted by the browser, all of its data, not parts of it, is deleted at the same time. If the origin had stored data by using IndexedDB and the Cache API for example, then both types of data are deleted.
```

| [Web Storage](/en-US/docs/Web/API/Web_Storage_API) | The Web Storage API provides mechanisms for webpages to store string-only key/value pairs, including [`localStorage`](/en-US/docs/Web/API/Window/localStorage) and [`sessionStorage`](/en-US/docs/Web/API/Window/sessionStorage). |
| [IndexedDB](/en-US/docs/Web/API/IndexedDB_API) | IndexedDB is a Web API for storing large data structures in the browser and indexing them for high-performance searching. |
| [Cache API](/en-US/docs/Web/API/Cache) | The Cache API provides a persistent storage mechanism for HTTP request and response object pairs that's used to make webpages load faster. |
| [Origin Private File System Access API (OPFS)](/en-US/docs/Web/API/File_System_Access_API#origin_private_file_system) | OPFS provides a file system that's private to the origin of the page and can be used to read and write directories and files. |

Note that, in addition to the above, browsers will store other types of data in the browser for an origin, such as [WebAssembly](/en-US/docs/WebAssembly) code caching.

> Browsers use different storage management systems for different types of data. For example, in most browsers today, Web Storage and cookies are not managed by the same system that's used to manage IndexedDB and Cache. Also note that that way browsers manage data changes over time. For example, in Firefox, Web Storage will soon start to use the same storage management system as the one used for IndexedDB.

Author: Good point, removing this section (which I also just realized now, was duplicated)

> Browsers use different storage management systems for different types of data. For example, Web Storage and cookies are not managed by the same system that's used to manage IndexedDB and Cache in most browsers. Also, note the way browsers manage data can change over time. For example, some browsers are changing how Web Storage is managed to reuse the same storage management system as IndexedDB.

Author: Yeah, I agree that we should remove this. It doesn't really help readers understand storage better, and will rot quickly with browsers implementation changing.

- Persistent: an origin can opt-in to store its data in a persistent way. Data stored this way is only evicted, or deleted, if the user chooses to, by using their browser's settings. To learn more, see [When is data evicted](#when-is-data-evicted).

The data stored in the browser by a origin is best-effort by default. When using web technologies such as IndexedDB or Cache, the data is stored transparently without asking for the user's permission. Similarly, when the browser needs to evict best-effort data, it does so without interrupting the user.

If, for any reason, developers need persistent storage (e.g., when building a web app that relies on critical data that isn't persisted anywhere else), they can do so by using the {{domxref("StorageManager.persist()")}} method of the {{domxref("Storage_API", "Storage API", "", "nocode")}}.

In Firefox, when a site chooses to use persistent storage, the user is notified with a UI popup that their permission is requested.

> Safari and most Chromium-based browsers, such as Chrome or Edge, automatically handle the permission request, and do not show any prompts to the user.

Author: Oh, that's great to know. I didn't have this understanding. I will use your phrasing.

Note that [research from the Chrome team](https://web.dev/persistent-storage/) shows that data is very rarely deleted by the browser. If a user visits a website regularly, there is very little chance that its stored data, even in best-effort mode, will get evicted by the browser.

### Private browsing

Note that in private browsing mode (also called _Incognito_ in Chrome, and _InPrivate_ in Edge), stored data does not persist and is usually deleted when the private browsing mode ends.

## How much data can be stored?

> > **Note:** The quotas documented below only apply to non private browser windows. In a private browser window, quotas are calculated differently and are usually much lower.

Author: I see. In general I don't think mentioning the inner workings of browsers is very useful, since they keep on changing and it can be super complex. But sometimes they can help understand a feature better, or they can explain a different behavior in some cases.

In this particular case, the way browsers implement storage in PB windows is probably not needed. But because some browsers have lower quotas in PB, it makes sense to keep a high-level note, to help readers understand why they're seeing a different behavior between PB and non-PB.

I'll rephrase this to make it more future-proof.

Web developers should wrap JavaScript that writes to browser storage within {{jsxref("Statements/try...catch","try...catch")}} blocks. Freeing up space by deleting data before storing new data is also recommended.

## When is data evicted?

Data eviction is the process by which a browser deletes an origin's stored data.

Data eviction can happen in multiple cases:

- When the device is running low on storage space, also known as _storage pressure_.
- When all of the data stored in the browser (across all origins) exceeds the total amount of space the browser is willing to use on the device.
- Proactively, for origins that aren't used regularly, which happens only in Safari.

### Storage pressure eviction

When a device is running low on storage space, also known as _storage pressure_, there may come a point when the browser has less available space than it needs to store all of the origin's stored data.

Browsers use a Least Recently Used (LRU) policy to deal with this scenario. The data from the least recently used origin is deleted. If storage pressure continues, the browser moves on to the second least recently used origin, and so on, until the problem is resolved.

> This eviction mechanism only applies to origins that are not persistent and skips over origins that have requested the user for persistent data by using {{domxref("StorageManager.persist()")}}).

Reviewer 3: WDYT about  something like: `.... and skips over origins that have been granted data persistence...`?
Only because Chromium and sounds like Safari as well, doesn't necessarily grant persistent when requested.

Author: Sounds great, making the change.

The amount of data browsers allow websites to store and the mechanisms they use to delete data when that limit is reached differs between browsers.

This article describes the web technologies that can be used to store data, the quotas that browsers have in place to limit websites from storing too much data, and the mechanisms they use to delete data when needed.

## How browsers separate data from different websites?

Browsers store the data from websites in different places to reduce risks of user being tracked across the web. In most cases, browsers manage stored data _per origin_.

The term _{{Glossary("origin")}}_ is therefore important to understand this article. An origin is defined by a scheme (such as HTTPS), a hostname, and a port. For example, `https://example.com` and `https://example.com/app/index.html` belong to the same origin because they have the same scheme (`https`), hostname (`example.com`), and default port.

The quotas and eviction criteria described in this article apply to an entire origin, even if this origin is used to run several websites, such as `https://example.com/site1/` and `https://example.com/site2/`.

> In some cases, however, browsers can decide to further separate the data stored by an origin in different partitions. For example, Chrome uses [Storage Partitioning](https://developer.chrome.com/en/docs/privacy-sandbox/storage-partitioning/) in cases where an origin is loaded within {{HTMLElement('iframe')}} elements in different third-party origins. For simplicity reasons, this article assumes that data is always stored per origin.

Author: @ayuishii I made a new section to explain the _per origin_ thing and added this paragraph about storage partitioning. Let me know if you think it's correct.

Reviewer 3: Thanks for the addition! LGTM

Reviewer 4: Note that Firefox also partitions third-party iframe's [storage keys](https://storage.spec.whatwg.org/#storage-keys) so that it is additionally keyed by the top-level origin.

(Also, at least for cookies, I believe Webkit has been double-keying/partitioning for an extremely long time.)

I do agree that the storage key seems is probably out of scope for this document.

### Cookies

Different browsers have different rules around how many cookies are allowed per origin and how much space these cookies can use on the disk. This article does not go into these details because cookies should not be used for storing data in the browser.

Different browsers have different rules around how many cookies are allowed per origin and how much space these cookies can use on the disk. While cookies are useful for preserving some small shared state between the browser and the web server across page navigation, using cookies for storing data in the browser is not advised. Cookies are sent with each and every HTTP request, so storing data in cookies that could be stored by using another web technology unnecessarily increases the size of requests.

Because cookies should not be used for storing data in the browser, cookie storage browser limits are not covered here.

### Web Storage

Web Storage, which can be accessed by using the {{domxref("Window.localStorage", "localStorage")}} and {{domxref("Window.sessionStorage", "sessionStorage")}} properties of the {{domxref("window")}} object, is limited to 10 MiB of data maximum on all browsers.

> Browsers can store up to 10 MiB of local storage, and 10 MiB of session storage per origin.

Reviewer 4: I think this is supposed to be 5 for both, not 10.  This is what you had in your [article](https://patrickbrosset.com/articles/2023-01-17-web-storage/) and what's defined for the endpoints [in the storage spec](https://storage.spec.whatwg.org/#storage-endpoints) and [internally for Firefox](https://searchfox.org/mozilla-central/rev/983c039377d4e97b381e13f4c3c1680360e6f3e2/modules/libpref/init/StaticPrefList.yaml#4019-4029).

Author: Yeah, that's also what I originally thought. I did a test where I was storing long strings of `a` characters to local storage and that test successfully showed that localStorage and sessionStorage were both limited to 5.

However, after talking to Nathan Memmott on the Chromium Slack, I realized that strings were stored in UTF-16.

I then did the same test with a string of `£` characters (2 bytes) and, there again, the test showed that I could only store 5*1024*1024 of them. Meaning that the quota was actually 10, not 5. And my `a` strings were just stored in UTF-16, where each `a` char also took 2 bytes.

This is at least what happens in Blink.

Reviewer 4: The check is the same in Firefox/Gecko.  But I think it's misleading to call that 10 MiB, because in Gecko (and it sounds like in Blink), our quota measurement and enforcement is on the string length of 5 * 1024 * 1024 code units across all keys and values, not on the actual disk usage.  Because JS uses 16-bit code units, this ends up being capable of storing 10 MiB of data.  But I think phrasing it "10 MiB" is going to lead to people making the wrong inference that they can put 10 * 1024 * 1024 code units worth of data in there.

I would suggest it's more appropriate to go into additional nuance here or to keep the 5 MiB number which is less likely to mislead.  Especially because it's easy for browsers to [accidentally](https://bugzilla.mozilla.org/show_bug.cgi?id=1681300) impose [USVString](https://mdn1.moz.one/en-US/docs/Web/API/USVString) semantics on LocalStorage values where illegal unicode sequences are replaced with U+fffd.  (Although the [Storage interface is defined in terms of DOMString](https://html.spec.whatwg.org/multipage/webstorage.html#the-storage-interface).)

Reviewer 5: Looking at it some more, it looks like it compares the [string length * 2](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/storage/storage_area_map.cc;l=11;drc=792356842a46bf11d6db111d205a151b3ab0b467;bpv=1;bpt=1) to [10 MiB](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/public/mojom/dom_storage/storage_area.mojom;l=53?q=third_party%2Fblink%2Fpublic%2Fmojom%2Fdom_storage%2Fstorage_area.mojom), so effectively 5 * 1024 * 1024 code units as @asutherland said. Still this means you can only store half the amount of unicode characters over U+10000, but this might not be an important detail.

Once this limit is reached, browsers throw a `QuotaExceededError` exception which should be handled by using a {{jsxref("Statements/try...catch","try...catch")}} block.

### Other web technologies

The data that's stored by using other web technologies, such as IndexedDB, Cache API, or Origin-Private File System Access API, is managed by a storage management system that's specific to each browser.

This system regulates all of the data that an origin stores using these APIs. When an origin's limit is reached, and the origin is best-effort, all of the data is evicted at the same time, no matter which of these APIs was used to store it.

Each browser determines, using whatever mechanism it chooses, the maximum amount of storage a given origin can use.

#### Firefox

> In Firefox, the maximum storage space an origin can use is calculated as 50% of the total disk space where the profile of the user is stored, capped at 8 TiB.

Reviewer 4: As [you correctly characterized in the related issue](https://github.com/mdn/content/issues/23557#issuecomment-1451690995), the 50% is only if the user granted persistent storage via navigator.storage.persist(), and 10% is the normal number.

A possibility of rephrasing this sentence and the next:

> In Firefox, the maximum storage space an origin can use in best-effort mode is calculated as 10% of the total disk space where the profile of the user is stored, capped at 8 TiB.  In persistent mode the percentage is raised to 50%.

> For example, if the device has a 500 GiB hard drive, Firefox will allow an origin in best-effort mode to store up to 50 GiB of data.  In persistent mode, 250 GiB of data is allowed to be stored.

Author: Oh that's right. Thanks for flagging this.
